src/OpenLoco/Audio/VehicleChannel.cpp

```diff
@@ -1,18 +1,94 @@
 #include "VehicleChannel.h"
 #include "../Entities/EntityManager.h"
 #include "../Interop/Interop.hpp"
+#include "../Map/TileManager.h"
+#include "../Ui/WindowManager.h"
 #include "../Vehicles/Vehicle.h"
 
 using namespace OpenLoco;
 using namespace OpenLoco::Audio;
 using namespace OpenLoco::Interop;
 
+uint8_t getZoomVolumeModifier(uint8_t zoom)
+{
+    switch (zoom)
+    {
+        case 0:
+            return 0;
+        case 1:
+            return 35;
+        default:
+            return 70;
+    }
+}
+
+uint8_t getUndergroundVolumeModifier(const Map::Pos3& pos)
+{
+    if (pos.x != Location::null && Map::validCoords(pos))
+    {
+        return 0;
+    }
+    auto* surface = Map::TileManager::get(pos).surface();
+    if (surface->baseZ() * 4 > pos.z)
+    {
+        return 28;
+    }
+    return 0;
+}
+
 static std::pair<SoundId, ChannelAttributes> sub_48A590(const Vehicles::Vehicle2or6* v)
 {
-    registers regs;
-    regs.esi = X86Pointer(v);
-    call(0x0048A590, regs);
-    return { static_cast<SoundId>(regs.eax), { regs.ecx, regs.edx, regs.ebx } };
+    auto* w = Ui::WindowManager::find(v->soundWindowType, v->soundWindowNumber);
+    auto* viewport = w->viewports[0];
+    const auto uiPoint = viewport->viewportToScreen({ v->sprite_left, v->sprite_top });
+
+    const auto zoomVolumeModifier = getZoomVolumeModifier(viewport->zoom);
+
+    const auto relativePositionX = (uiPoint.x << 16) / std::max(Ui::width(), 64);
+    const auto panX = (relativePositionX - (1 << 15)) / 16;
+    const auto relativePositionY = (uiPoint.y << 16) / std::max(Ui::height(), 64);
+    const auto panY = (relativePositionY - (1 << 15)) / 16;
```

Comment:
Consider moving this into a function like ```std::pair<int32_t, int32_t> calculcatePan(...)```

Reply:
yeah i already had but for some reason net connection went down when i uploaded :(